pallets/system/src/header_builder.rs

```diff
@@ -263,3 +264,207 @@ impl Decode for AvailExtrinsic {
 		})
 	}
 }
+
+fn data_root_avail(app_ext: Vec<AppExtrinsic>) -> [u8; 32] {
+	let ext: Vec<Vec<u8>> = app_ext.iter().map(|x| x.data.clone()).collect();
+	let avail_ext = ext
+		.iter()
+		.filter_map(|e| AvailExtrinsic::decode(&mut &e[..]).ok())
+		.collect::<Vec<_>>();
+	let data_root: [u8; 32] = if avail_ext.len() > 0 {
+		let leaves: Vec<[u8; 32]> = avail_ext.iter().map(|e| Sha256::hash(&e.data)).collect();
+		let tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+		tree.root().expect("Data Root computation failed")
+	} else {
+		Default::default()
+	};
+	data_root
+}
+
+mod tests {
+	use super::*;
+
+	/// tests for data root is created correctly
+	#[test]
+	fn data_root_test() {
+		let avail_ext: Vec<AppExtrinsic> = vec![
+			AppExtrinsic {
+				app_id: 0,
+				data: vec![40, 4, 3, 0, 11, 194, 98, 8, 55, 131, 1],
+			},
+			AppExtrinsic {
+				app_id: 3,
+				data: vec![
+					93, 2, 132, 0, 28, 189, 45, 67, 83, 10, 68, 112, 90, 208, 136, 175, 49, 62, 24,
+					248, 11, 83, 239, 22, 179, 97, 119, 205, 75, 119, 184, 70, 242, 165, 240, 124,
+					1, 196, 71, 85, 121, 78, 169, 73, 233, 65, 3, 144, 203, 76, 224, 127, 226, 216,
+					6, 134, 86, 24, 91, 90, 185, 180, 62, 239, 147, 76, 54, 128, 71, 137, 104, 193,
+					248, 62, 54, 10, 93, 148, 47, 231, 94, 157, 88, 228, 145, 6, 168, 232, 178, 54,
+					1, 203, 198, 166, 51, 216, 14, 93, 8, 157, 131, 164, 0, 4, 0, 3, 0, 0, 0, 29,
+					1, 164, 104, 104, 97, 106, 107, 97, 110, 99, 107, 97, 32, 108, 97, 104, 32,
+					105, 97, 107, 106, 32, 99, 97, 105, 117, 104, 32, 97, 98, 32, 97, 105, 97, 106,
+					104, 32, 97, 32, 103, 97, 104, 97,
+				],
+			},
+		];
+		let empty_data: Vec<AppExtrinsic> = vec![];
+		let single_data: Vec<AppExtrinsic> = vec![AppExtrinsic {
```

Comment:
so this extrinsic is not a `submitData()` call?

Reply:
yes it is a submitdata() but here only the decoded data is directly given into the function

Reply:
Then how is the root empty in this case?

Reply:
yes